File: src/prod/src/managed/Api/src/System/Fabric/Description/NetworkNodeQueryDescription.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace System.Fabric.Description
{    
    using System;
    using System.Fabric.Common;
    using System.Fabric.Interop;

    /// <summary>
    /// <para>Represents the query input used by <see cref="System.Fabric.FabricClient.NetworkManagementClient.GetNetworkNodeListAsync(NetworkNodeQueryDescription)" />.</para>
    /// </summary> 
    public sealed class NetworkNodeQueryDescription : PagedQueryDescriptionBase
    {
        /// <summary>
        /// <para>Initializes a new instance of the <see cref="System.Fabric.Description.NetworkNodeQueryDescription" /> class.</para>
        /// </summary>
        public NetworkNodeQueryDescription()
        {
            this.NetworkName = null;
            this.NodeNameFilter = null;
            this.ContinuationToken = null;
        }

        /// <summary>
        /// <para>Gets or sets name of the network to query for.</para>
        /// </summary>
        public string NetworkName { get; set; }

        /// <summary>
        /// Gets or sets the node name used to filter the nodes to query for.
        /// The node that is of this node name will be returned.
        /// </summary>
        public string NodeNameFilter { get; set; }

        internal static void Validate(NetworkNodeQueryDescription queryDescription)
        {
            Requires.Argument<string>("NetworkName", queryDescription.NetworkName).NotNullOrWhiteSpace();

            ValidatePaging(queryDescription);
        }

        internal IntPtr ToNative(PinCollection pinCollection)
        {
            var nativeDescription = new NativeTypes.FABRIC_NETWORK_NODE_QUERY_DESCRIPTION();
            nativeDescription.NetworkName = pinCollection.AddObject(this.NetworkName);
            nativeDescription.NodeNameFilter = pinCollection.AddObject(this.NodeNameFilter);
            nativeDescription.PagingDescription = this.ToNativePagingDescription(pinCollection);
            nativeDescription.Reserved = IntPtr.Zero;

            return pinCollection.AddBlittable(nativeDescription);
        }
    }
}

File: src/prod/src/ServiceModel/NetworkNodeQueryDescription.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace ServiceModel;
using namespace std;
using namespace Common;
using namespace Query;

StringLiteral const TraceComponent("NetworkNodeQueryDescription");

NetworkNodeQueryDescription::NetworkNodeQueryDescription()
    : networkName_()
    , nodeNameFilter_()
    , queryPagingDescription_()
{
}

ErrorCode NetworkNodeQueryDescription::FromPublicApi(FABRIC_NETWORK_NODE_QUERY_DESCRIPTION const &networkNodeQueryDescription)
{
    TRY_PARSE_PUBLIC_STRING(networkNodeQueryDescription.NetworkName, networkName_);

    TRY_PARSE_PUBLIC_STRING_ALLOW_NULL(networkNodeQueryDescription.NodeNameFilter, nodeNameFilter_);
    
    if (networkNodeQueryDescription.PagingDescription != nullptr)
    {
        QueryPagingDescription pagingDescription;
        
        ErrorCode error = pagingDescription.FromPublicApi(*(networkNodeQueryDescription.PagingDescription));
        if (!error.IsSuccess())
        {
            Trace.WriteInfo(TraceComponent, "PagingDescription::FromPublicApi error: {0}", error);
            return error;
        }

        queryPagingDescription_ = make_unique<QueryPagingDescription>(move(pagingDescription));
    }

    return ErrorCode::Success();
}

void NetworkNodeQueryDescription::GetQueryArgumentMap(__out QueryArgumentMap & argMap) const
{
    // NetworkName is a mandatory query parameter.
    argMap.Insert(
       QueryResourceProperties::Network::NetworkName,
       networkName_);

    if (!nodeNameFilter_.empty())
    {
        argMap.Insert(
            Query::QueryResourceProperties::Node::Name,
            nodeNameFilter_);
    }

    if (queryPagingDescription_ != nullptr)
    {
        queryPagingDescription_->SetQueryArguments(argMap);
    }
}
